File: challenges/ttl-store/go/solution/exercise.go
package exercise

import "refresher/tooling/go/contracts"

type entry struct {
	value    string
	deadline int64
}

type TTLStore struct {
	clock   contracts.Clock
	entries map[string]entry
}

func NewTTLStore(clock contracts.Clock) *TTLStore {
	return &TTLStore{clock: clock, entries: make(map[string]entry)}
}

func (s *TTLStore) Put(key, value string, ttlMS int64) {
	s.entries[key] = entry{value, s.clock() + ttlMS}
}

func (s *TTLStore) Get(key string) (string, bool) {
	e, exists := s.entries[key]
	if !exists {
		return "", false
	}
	if s.clock() >= e.deadline {
		delete(s.entries, key)
		return "", false
	}
	return e.value, true
}

func (s *TTLStore) Delete(key string) bool {
	if _, exists := s.Get(key); !exists {
		return false
	}
	delete(s.entries, key)
	return true
}

File: challenges/ttl-store/typescript/solution/exercise.ts
import type { Clock, KeyValueStore } from "../../../../tooling/typescript/contracts.js";

export class TTLStore implements KeyValueStore {
  private entries = new Map<string, { value: string; deadline: number }>();

  constructor(private clock: Clock) {}

  put(key: string, value: string, ttlMs: number): void {
    this.entries.set(key, { value, deadline: this.clock() + ttlMs });
  }

  get(key: string): string | null {
    const entry = this.entries.get(key);
    if (entry === undefined) return null;
    if (this.clock() >= entry.deadline) {
      this.entries.delete(key);
      return null;
    }
    return entry.value;
  }

  delete(key: string): boolean {
    if (this.get(key) === null) return false;
    return this.entries.delete(key);
  }
}
